### DISCUSSION TITLE
Documenting the "power utilization" features of Netbox

### DISCUSSION THREAD

**Discussion by candlerb** [netbox-community/netbox#12837] [General]
There seems to be very little documentation about the power utilization features and calculation in Netbox. [Here](https://docs.netbox.dev/en/stable/models/dcim/powerfeed/#max-utilization) it says:

> The power utilization of a rack is calculated when one or more power feeds are assigned to the rack and connected to devices that draw power.

And indeed, I see a power utilization figure at rack level. But it doesn't say how it relates to "Maximum draw" and "Allocated draw" on a power port; nor how the "Max utilization (percentage)" on a power feed relates to either "Maximum draw" or "Allocated draw"; nor how you should model devices with multiple PSUs (e.g. if the device average draw is 70W, should you allocate 35W on PSU1 and 35W on PSU2? Of course, if one PSU or feed fails, then you'll get 70W via the other one)

Therefore, I'm attempting to reverse engineer the logic. I'm posting it here so people can comment and/or correct it.

#### Power Ports

* A Power Port on a Device has attributes "maximum_draw" and "allocated_draw" which are integer numbers of Watts, with a minimum of 1 but may be null (`netbox/dcim/models/device_components.py`)
* A Power Port can be connected, via a cable, to a single Power Outlet or Power Feed (COMPATIBLE_TERMINATION_TYPES in `netbox/dcim/constants.py`)
* Within a PDU-type device, one or more Power Outlets can internally connect to a shared Power Port (inlet)
* There are some internal helper functions on class PowerPort in `netbox/dcim/models/device_components.py`:
    * `get_downstream_powerports`: on a PDU it finds the PowerOutlets internally linked to this PowerPort, and then any downstream PowerPorts connected by cables. But AFAICS it only looks one level deep, i.e. it doesn't recurse through further PDUs
    * `get_power_draw` performs a summary calculation on a (PDU) power port
        * if the power port has either maximum_draw or allocated_draw statically assigned, then it returns both these values (and if one of these values is null then it is converted to 0)
        * otherwise, it finds all the ports from `get_downstream_powerports` and sums the maximum_draw and allocated_draw from these. It uses a django.db `aggregate()` with two `Sum()`, so I think this is done using the values stored in SQL only (I need to test, if any value is null then will the entire result be null?)
        * in addition, if this power port is connected to a power feed, and that power feed is marked as three-phase, then the calculation is repeated for each phase (leg), filtering the set of downstream power ports to those on that given leg

At this point, we can set up some simple test cases. I will ignore 3-phase for now. To start:

* Create a test site and a test rack
* Create a device type "8-port PDU" with 1 power port (`IN`) and 8 linked power outlets (`OUT[1-8]`)
* Create instance "TestPDU" of this type
* Create device "Test1" with a power port maximum_draw=100, allocated_draw=30, and cable to OUT1
* Create device "Test2" with a power port maximum_draw=90, allocated_draw=40, and cable to OUT2
* Create device "Test3" with a power port maximum_draw=80, allocated_draw unset (null), and cable to OUT3
* Create device "TestPDU2" and connect its IN to TestPDU OUT4
* Create device "Test4" with power port maximum_draw=70, allocated_draw=50, and connect to TestPDU2 port OUT1

```
TestPDU
,--------.
|   ,OUT1|--- Test1 (100,30)
|   ,OUT2|--- Test2 (90, 40)       TestPDU2
|IN--OUT3|--- Test3 (80, null)    ,--------.
|   `OUT4|------------------------|IN--OUT1|--- Test4 (70,50)
|   `... |                        `--------'
`--------'
```

Now look at the "TestPDU":

![image](https://github.com/netbox-community/netbox/assets/44789/fdee242b-9acd-470d-875b-c9dce81619ac)

And "TestPDU2":

![image](https://github.com/netbox-community/netbox/assets/44789/e8993866-2c0b-4627-8949-1600eb4c583c)

Hence on TestPDU:

* "Allocated" is the sum of "allocated_draw" on the directly-connected devices (30+40)
* The null allocated_draw on Test3 doesn't affect the calculation
* The grandchild device on TestPDU2 is *not* counted
* The units are inconsistent (VA or W; VA is Apparent Power whilst W is Real Power). TBH this isn't too important, as any half-decent data centre equipment should have a power factor close to 1; however it would be nice to be consistent
* The total maximum_draw isn't displayed

What about the "Available" and "Utilization" columns? These are inserted in the template `netbox/templates/dcim/device.html`, and are only displayed if the powerport is connected to a powerfeed (`powerfeed=powerport.connected_endpoints.0`, and additionally this must have an attribute `available_power`, which is only an attribute of powerfeed not a power outlet)

If so, the two columns displayed are

* `powerfeed.available_power`
* `{% utilization_graph utilization.allocated|percentage:powerfeed.available_power %}`

(also if there are three-phase legs, additional table rows are displayed for each leg)

Therefore, next we need to go into power feeds.

#### Power Feeds

* A power panel *must* be associated with a Site
* A power panel *may* be associated with a Location (but not a Rack)
* A power feed *must* be associated with a power panel (and hence implicitly with a Site and possibly Location)
* A power feed *may* be associated with a Rack
* A power feed includes the following attributes (`netbox/dcim/models/power.py`, `netbox/dcim/choices.py`):
    * type (primary, redundant)
    * supply (ac, dc)
    * phase (single-phase, three-phase)
    * voltage (integer, signed, non-zero)
        * there is a check that the voltage cannot be negative for AC supply
        * but there is no check that the phase cannot be "three-phase" for a DC supply
    * amperage (integer, minimum 1)
    * max_utilization (integer between 1 and 100)
    * available_power (integer, non-editable, calculated on save)
        * it is calculated as `abs(self.voltage) * self.amperage * (self.max_utilization / 100)` and then multiplied by 1.732 if three-phase; finally rounded to nearest integer
        * again, it is described as "power" but is actually volt-amps

A Power Port can be connected with a cable to a Power Feed, and the usage calculations on the Power Port take into account attributes of the powerfeed.

Let's extend the test data:

* Create a Power Panel "test-panel" in our test site
* Create a Power Feed "test-feed" in our test rack (AC, 230V, 15A, single phase, 80% max util)
* Connect to "TESTPDU" power port "IN"

Now we see on TESTPDU:

![image](https://github.com/netbox-community/netbox/assets/44789/c1808180-cb06-4cbd-98e4-f608d21c23dd)

This has calculated `230*15*0.8 = 2760 VA` as the maximum utilization, and the sum of devices connected to directly-connected PDUs as 70VA, hence 2.5% utilization (rounded from 2.536%)

We see the same when looking at the Power Feed itself:

![image](https://github.com/netbox-community/netbox/assets/44789/c35bbc68-7a17-4a94-a196-2996d2bc5628)

#### Rack-level calculations

If the test-feed is associated with the test-rack, then the test-rack shows some power utilization:

![image](https://github.com/netbox-community/netbox/assets/44789/41acdf97-60e4-4e2c-a3ac-802ee4809d37)

But weirdly, it shows utilization at 2.0% not 2.5%.  It is calculated slightly differently, in function `get_power_utilization` in class Rack (`netbox/dcim/models/racks.py`):

* It finds all power feeds linked to this rack
* It calculates available_power_total = sum(pf.available_power for pf in powerfeeds)
* For all powerfeeds, it finds all the link_peers of type `PowerPort`
* It then sums `get_power_draw()['allocated']`, divides by available_power_total, multiplies by 100, and rounds *down* to an integer:
    * `return int(allocated_draw / available_power_total * 100)`

I think this is a bug: the display of "2.0%" implies that it has at least 1 decimal place of accuracy.  In the other places, the percentage is calculated using percentage() in `netbox/utilities/templatetags/helpers.py` which rounds to nearest 0.1%:

```
return round(x / y * 100, 1)
```

I will raise this separately (#12838).

If the test-feed is not associated with the rack, then the rack shows 0% power used:

![image](https://github.com/netbox-community/netbox/assets/44789/baa2d523-2b41-4353-a9e4-097b98407db1)

(In principle, it could still sum the power port usage of all devices within that rack, even if they are connected to power feeds in other racks, or even if they are not connected to any power feed - but it doesn't)

As far as I can see, there is no power information displayed at Site or Location level.  Therefore, any power feed which is not associated with a rack, does not have its usage counted, aggregated or displayed, except when looking at the Power Feed itself.  (Even when you view a Power Panel, you see a list of the Power Feeds it contains, but no aggregate power information)

#### Best practices?

The question then arises as to how you should properly make use of these features, especially how to assign values to "allocated_draw" and "maximum_draw".

In general, the questions I want to answer from the data fall into two classes:

* "How much power am I using in this rack, or in this site?"  This is to compare usage with data centre contracts.  In my experience, data centres don't care at the level of power feeds: if you've contracted for 2.5kW in a given rack, they don't enforce that it's 1.25kW on A and 1.25kW on B.  Indeed, often they will only consider the aggregate usage in a particular site, or location within the site (e.g. cage). That's because the only thing that they pay for is the total power you use, regardless of where it goes.
* "How close am I to blowing the fuse on each PDU?"  In this case, you care more about current than power.  Part of this is about in-rush current if the whole rack were powered up at once, although that's a rare event, and you could probably power up in phases if necessary.  More important is what happens if PDU A is knocked out and all the devices fall over to being powered from PDU B (or vice versa).

Specifically then, if you have a device with two power ports, which uses a nominal 500W, how should you model it?

* Set allocated_draw to 250W and maximum_draw to 500W on both power ports.
    * ✅ You count all power feeds allocated_draw towards total site/rack usage (which Netbox does)
    * ✅ It accurately reflects the normal usage on each PDU
    * ✅ It correctly assigns usage for other devices with a single PSU to the PDU that they connect to
    * ❌ The total *maximum* draw on a PDU is not reported within Netbox (AFAICS), so you don't know the worst-case current draw when everything fails over to a single PDU - but you could make a custom report
    * ❌ You have to manually apportion the power between PSUs, and this changes if you add a second PSU to a device
* Set allocated_draw to 500W on first power port and 0W on second. Connect PSU A to PDU A on "primary" power feed and PSU B to PDU B on "redundant" power feed.
    * ✅ It *does* give the peak utilization on PDU A when PDU B is out of service (but not vice versa)
    * ❌ It isn't exactly accurate; in reality, normally the device will draw 250W from each
* Set allocated_draw to 500W on both power ports. Connect PSU A to PDU A on "primary" power feed and PSU B to PDU B on "redundant" power feed. Then don't count the "redundant" power feed towards the data centre contracts
    * ✅ It *does* give the peak utilization when either PDU is out of service
    * ❌ AFAICS, netbox doesn't take into account "primary" or "redundant" when working out total rack power utilization; power will be double-counted
    * ❌ It isn't exactly accurate; in reality, normally the device will draw 250W from each
    * ❌ You may have some single-PSU devices in the rack, and connect some to PDU A and others to PDU B. This would then fail to count the usage of the "B" devices towards the total

(The above assumes that we're not interested in in-rush current at all, and rather that "maximum_draw" is intended to reflect the draw when a device is running on the minimum number of PSUs, after a failure of a PSU or a PDU feed)

Then the issue is how to report on this.

Firstly, since Netbox doesn't handle second-level PDUs it makes the power-feed level accounting unusable in that situation. Specific examples:

* if I run out of power ports on a PDU and connect a Y-splitter, I model that as another device
* if I connect a G-type power strip into an IEC PDU, that is also modelled as another device

Secondly, there's no site-level reporting, which I would find the most useful.

AFAICS, the best approach for now is to write some custom reports:

* One for Site and/or Location level
    * For all Devices in that Site/Location, sum all the Power Port allocated_draw figures (possibly ignoring Power Ports which are connected internally to Power Outlets in the same device)
    * To get utilization as a percentage of available_power, sum the Power Feeds in all Power Panels in that Site and/or Location
* One at Rack level
    * Do the same for all Devices in that Rack
    * To get utilization as a percentage of available_power, sum the Power Panels assigned to that Rack (although this won't work if power is drawn from a Power Feed which serves the Location or Site but not linked to that Rack)
* To look at max utilization on each power feed, sum all the maximum_utilization figures for all power ports connected directly or indirectly to that power feed
    * Configure max utilization on each power port as the worst-case value if the device were running on the *minimum* number of power ports (e.g. if a device has four PSUs and it can run on three, and uses 600W total, then it would be 150W allocated and 200W maximum on each PSU)
* Arguably, the reports should exclude devices whose status is "Offline" (although I'd probably include "Failed", on the basis that they are likely to be fixed quickly)

If people have other experience of this or better suggestions for how to use it, I'd be very happy to hear them.

**Comment by mtinberg** [comment]
Set allocated_draw to 500W on both power ports. Connect PSU A to PDU A on "primary" power feed and PSU B to PDU B on "redundant" power feed. Then don't count the "redundant" power feed towards the data centre contracts
✅ It does give the peak utilization when either PDU is out of service
❌ AFAICS, netbox doesn't take into account "primary" or "redundant" when working out total rack power utilization; power will be double-counted
❌ It isn't exactly accurate; in reality, normally the device will draw 250W from each
❌ You may have some single-PSU devices in the rack, and connect some to PDU A and others to PDU B. This would then fail to count the usage of the "B" devices towards the total

While there is a list of possible enhancements in your detailed document, I think there is a missing column in the data for storing an actual​ utilization number (scraped via external process over SNMP, API, etc.) and a need to mark and elide redundant data when aggregating at a higher level.  I think you are right in that the way this should work is the max should be the rating for the power supply (1100W or whatever) the allocation should be how much power needs to be reserved to run the device off this sole power supply (500W) and an actual value (maybe just a custom_field for those who want it) to show measured utilization (250W).  I am remined of PoE classes where the edge device might draw 5w but need to be allocated 15w based on its class and the switch needs to reserve enough for the total draw possible or it'll refuse to enable more.  In this case its about modeling allocation for the PDUs so you don't exceed 50% of their capacity and they can provide useful redundancy of a circuit outage.

so add a flag for redundant connections so they don't get double-counted at the power-panel level and add a custom_field for caching actual current data, to cross-check your allocations.  Then you can see actual, allocated and max possible utilization at the power panel level.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison

**Reply by candlerb** [reply]
The trouble with automatically ingesting measured power draw is:

1. It can vary widely from second to second, e.g. depending on CPU active/idle
2. It goes against the Netbox philosophy of "source of truth, not a monitoring system"
3. At the device level, it only works if you have intelligent PDUs.  (In most places, I only get total power from the PDU, not per socket)

If you *could* get per-port figures, then it would be reasonable to monitor the *average* draw over a week or month, and use that to guide what the "allocated draw" is configured as.

In reality, I think the *measured* draw has much more concrete value anyway, as that's what the data centre will charge on - and it's likely to be an aggregate (per PDU / power feed), rather than at the level of individual devices.

The value of the per-device power draw modelling is that it can *predict* draw when planning new capacity.  If you can create a set of per-device-type allocations which sum to roughly the right amount (i.e. match the measured value for existing kit), then you can get an idea how many new devices you could add before you bust your power budget.

Manufacturer specs are likely to err on the high side, so I think you're right that to be useful, these power draw figures should be representative of real devices in real service.

> I think you are right in that the way this should work is the max should be the rating for the power supply (1100W or whatever)

I don't think that's what I was suggesting; the power supply *rating* is not something I had considered.  I guess it could be of interest, but IMO it probably belongs in the module type or inventory item type, since it's not something you can adjust - it's just part of the device specification.  I don't think it is very meaningful to sum it across all devices.  It's very much a worst-case figure: if you stuffed a single server with the maximum number of GPU cards or hard drives or whatever, whether the PSU itself could cope.

> the allocation should be how much power needs to be reserved to run the device off this sole power supply (500W)

However, Netbox's current summing of these will show the device using 500+500=1000W, when it's only using 500W, so as you say, you'd need to exclude the double-counted values somehow.  I don't think marking a particular power feed as "redundant" is ideal, because you may have some devices that are not dual-powered.  For example, I sometimes have a redundant pair of routers each with single PSU, and connect one to the A feed and one to the B.

I am inclined to say that allocated/expected power draw should be an attribute at the Device level, not at the Power Port level, because in general devices don't have active / standby power supplies, but instead load balance between them. But to model that today in Netbox you need to divide the draw between the power ports.

If you take that view that "allocated draw" is the share of power that the port draws in normal operation, then I think "maximum power draw" should be the worst-case amount in a failure scenario.  If the server has 2 power supplies but can run off 1, then it should be the full draw.  If the server has 3 power supplies and can run off 2 (but not 1), then it should be half the full draw.

> In this case its about modeling allocation for the PDUs so you don't exceed 50% of their capacity and they can provide useful redundancy of a circuit outage.

Exactly: I think that's what I think the "maximum draw" should be aiming to model.  "If we lost a PDU, would the second PDU have enough capacity to cope?"

**Comment by mtinberg** [comment]
The trouble with automatically ingesting measured power draw is:
Sure, this probably doesn't belong exactly in core but having a plugin or custom_field to scrape this data so you can report on discrepancies between the model and reality (kind of like how the napalm/lldp audit was moved to a plugin) seems useful.  I think OpenDCIM does something like that, but you are right it probably doesn't belong in core.

I don't think that's what I was suggesting; the power supply rating is not something I had considered. I guess it could be of interest, but IMO it probably belongs in the module type or inventory item type, since it's not something you can adjust - it's just part of the device specification. I don't think it is very meaningful to sum it across all devices. It's very much a worst-case figure: if you stuffed a single server with the maximum number of GPU cards or hard drives or whatever, whether the PSU itself could cope.


I was thinking (not clearly as I haven't tackled this problem seriously yet, its on the backlog somewhere) of needing to know the power supply capacity more for tracking PoE switch utilization as that's the more pressing issue for my site, what's the max draw/heat load in a room and what would happen if someone filled the access stack with a bunch of PoE devices (cameras, WAPs, cardaccess, phones, etc.).  Would we need to swap out power supplies (750W with 1100W) on how many stack members and would we need facilities to pull new electric circuits or cooling to a room.


However, Netbox's current summing of these will show the device using 500+500=1000W, when it's only using 500W, so as you say, you'd need to exclude the double-counted values somehow. I don't think marking a particular power feed as "redundant" is ideal, because you may have some devices that are not dual-powered. For example, I sometimes have a redundant pair of routers each with single PSU, and connect one to the A feed and one to the B.

I was thinking of marking the power supply/power port as redundant or not rather than the power outlet or feed, and it looks like the current model has redundancy as part of the feed, which is great except for single-homed devices who's only power comes from the "redundant" feed, as you pointed out.  Redundancy happens at the port->outlet level a lot, I suppose it happens at the feed level if you use an ATS in front of the PDU.  That's it, get an ATS and another PDU and then you can put your single-homed devices behind a dual-homed PDU behind a primary/redundant PDU behind the power feed and then you just need to solve the recursive lookup problem (kidding).

—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________

**Reply by candlerb** [reply]
Thanks. There are plenty of possibilities for expanding the current model.

The point I started from was "how are the *existing* power features supposed to be used?" and it would be good to see some examples of how they are being used in practice, if anyone is.

**Comment by pobradovic08** [comment]
Our goal when modeling power in DCIM is not to see how much power are we using, but to see how close are we to 50% utilization of each PDU fuse.

How this looks in practice for us:

For example we have 230V PDUs A and B in a rack. Each PDU has two groups with it's own fuse (16A).

- We model each 16A fuse as a separate power feed (to see utilization per fuse, not per PDU)
- Each feed has its maximum utilization set to 50% (8A, as we want to maintain redundancy in case of PDU failure and never cross the 8A mark under normal operation)
- If we can measure peak device power draw to be 500W, allocated power per PSU is 500W devided by the number of PSUs. For a device with 5 PSUs, that would be 100W allocated per PSU.
- For maximum PSU power we set 500W. If a device has 5 PSUs and can function with minimum of 2, maximum power per PSU is 500W divided by minimum required number of PSUs (in this case that would be 500W/2=250W). Just as you described in one of the comments.

The current system kinda works but it's a bit hard to figure out and clumsy to setup.

I would also like to have a power utilization set on the device level, and then to distribute the load evenly across all present power supplies by default. For devices that don't have even distribution of a load, we can have a factor setting for each PSU that would control how much of the power should be distributed to that PSU. My biggest issue with the current system is that there are devices which power consumption is highly dependent on the populated bays/modules. For example blade chassis that can hold 16 server blades will have different power usage with 1 blade and all 16 blades. This means that I have to change all of the PSUs every time we add/remove a blade.

I think that it would be great to have power consumption for each bay/module (alongside the parent device), aggregate all of that on a device level and then distribute the aggregate value across available power supplies.

**Comment by candlerb** [comment]
FYI, I've uploaded a custom script [here](https://github.com/netbox-community/customizations/pull/96/files) which may be useful when considering power draw at a site level (as opposed to current at the PDU level)

It calculates the total power draw across all devices in a site, just by summing allocated power from each device's power port in that site. It doesn't make any use of power feeds, and doesn't use the parent PDU power calculation. In fact, you don't even have to connect any cables to the power ports.

It can also count the total power outlets in a site, and how many of those are available (not connected).

**Comment by hogduske** [comment]
I don't know if this is the right place for my question.

As mentioned in the original post 
_"The grandchild device on TestPDU2 is not counted"_
Why is that so and is there a way to change that behaviour.

In our use case we have four power feeds coming into our rack. Two primary and two redudant ones.
Each powerfeed is then hardwired into a 14 socket C14 PDU which then feeds multiple "sub-PDUs".
The deviced are connected to the sub-PDUs.
AFAIK there is no way to report the allocated draw of the devices all the way to the powerfeed.
Rack power utilization always stays at 0%.

Am I doing something wrong or is this how its intended to be?

**Reply by candlerb** [reply]
> is there a way to change that behaviour.

Anything is possible in software. It's just a question of having the resource and focus to design and implement a new approach for doing the calculation.

Part of the problem is defining *exactly* what that behaviour should be (and indeed, defining exactly how that differs from the *current* behaviour, which isn't really documented - hence the point of this thread).

> AFAIK there is no way to report the allocated draw of the devices all the way to the powerfeed

You can do it using a custom report, as I linked to above. (Not tested with or adapted for Netbox 4)

**Comment by ZPrimed** [comment]
I don't have a ton to add here as the existing discussion is pretty thorough, but...

There definitely _are_ devices with dual-PSUs that are not load-sharing. At least, that's the behavior I observe on Mikrotik hardware - it only really draws from one supply at a time. This is very obvious when you are using DC PSUs attached to something that can monitor current on each output.

The existing system in NetBox is very AC-focused, which is understandable given its origins. It's somewhat limiting for a company that only uses AC up to an intelligent rectifier system, and then powers everything in the rack with DC from that rectifier... but this is probably a bit off-topic for the discussion here so I'm not going to draw that out anymore.

**Reply by FlorianHeigl** [reply]
Many Lenovo servers also don't normally run in load sharing redudant mode and don't have a power policy setting for it like Dell has.
The Bull ones I got also. 
The opposite problem also exists, how to model a combined ATS+PDU inside the rack.
